File: android/app/src/main/kotlin/io/mynote/app/sync/NoteRepository.kt
package io.mynote.app.sync

import io.mynote.app.data.MyNoteDatabase
import io.mynote.core.Block
import io.mynote.core.BlockContent
import io.mynote.core.BlockType
import io.mynote.core.Change
import io.mynote.core.Entity
import io.mynote.core.FractionalIndex
import io.mynote.core.InlineSpans
import io.mynote.core.Note
import io.mynote.core.WelcomeNote
import io.mynote.core.ThemeSpec
import io.mynote.core.asChange
import kotlinx.serialization.json.JsonPrimitive
import java.util.UUID

/**
 * Every write the UI makes goes through here.
 *
 * One job: stamp the edit with a fresh clock, persist it locally, queue it for
 * sync, and return — without awaiting the network. Callers treat these as
 * instant, because locally they are.
 */
class NoteRepository(
    private val db: MyNoteDatabase,
    private val coordinator: SyncCoordinator,
) {
    suspend fun createNote(title: String = "", parentId: String? = null, after: String? = null): String {
        val id = UUID.randomUUID().toString()
        write(
            Note(
                id = id,
                title = title,
                parentId = parentId,
                orderKey = FractionalIndex.between(after, null),
                hlc = coordinator.engine.stamp(),
            ).asChange()
        )
        // A new page opens with one empty paragraph, so the cursor has somewhere
        // to land instead of an empty screen.
        appendBlock(id, null)
        return id
    }

    suspend fun rename(noteId: String, title: String, icon: String?, parentId: String?, orderKey: String) {
        write(Note(noteId, title, icon, parentId, orderKey, coordinator.engine.stamp()).asChange())
    }

    suspend fun deleteNote(noteId: String) {
        val blockIds = db.blocks().idsForNote(noteId)
        write(Change(Entity.NOTE, noteId, coordinator.engine.stamp(), deleted = true))
        // Tombstone the blocks too, or they would linger on other devices.
        for (blockId in blockIds) {
            write(Change(Entity.BLOCK, blockId, coordinator.engine.stamp(), deleted = true))
        }
    }

    suspend fun appendBlock(
        noteId: String,
        after: String?,
        before: String? = null,
        type: BlockType = BlockType.PARAGRAPH,
    ): String {
        val block = Block(
            id = UUID.randomUUID().toString(),
            noteId = noteId,
            orderKey = FractionalIndex.between(after, before),
            type = type,
            hlc = coordinator.engine.stamp(),
        )
        write(block.asChange())
        return block.id
    }

    /**
     * Return pressed: the text before the caret stays, the rest becomes a new
     * block underneath.
     *
     * @return the id of the new block, so the caller can move the caret into it.
     */
    suspend fun splitBlock(
        block: Block,
        before: String,
        after: String,
        nextOrderKey: String?,
    ): String {
        // Formatting goes with the words. The offsets come from the original
        // text rather than from `before` and `after` alone, because a Return
        // pressed over a selection deletes the middle — so the tail does not
        // start where the head ends.
        val originalText = block.content.text
        val originalSpans = block.content.inlineSpans
        val tailStart = originalText.length - after.length

        update(
            block.copy(
                content = block.content.copy(text = before)
                    .withSpans(InlineSpans.slice(originalSpans, originalText.length, 0, before.length))
            )
        )

        val tail = Block(
            id = UUID.randomUUID().toString(),
            noteId = block.noteId,
            orderKey = FractionalIndex.between(block.orderKey, nextOrderKey),
            // A list carries on as a list; a heading does not, because the line
            // after a heading is almost never another heading.
            type = if (block.type.continuesOnSplit) block.type else BlockType.PARAGRAPH,
            content = BlockContent(text = after).withSpans(
                InlineSpans.slice(originalSpans, originalText.length, tailStart, originalText.length)
            ),
            hlc = coordinator.engine.stamp(),
        )
        write(tail.asChange())
        return tail.id
    }

    /**
     * Backspace at the start of a block: fold it into the one above.
     *
     * @return the caret offset in the previous block, i.e. the join point.
     */
    suspend fun mergeIntoPrevious(block: Block, previous: Block): Int {
        val joinOffset = previous.content.text.length

        if (block.content.text.isNotEmpty()) {
            val joined = previous.content.text + block.content.text
            update(
                previous.copy(
                    content = previous.content.copy(text = joined).withSpans(
                        InlineSpans.concatenated(
                            previous.content.inlineSpans, previous.content.text.length,
                            block.content.inlineSpans, block.content.text.length,
                        )
                    )
                )
            )
        }
        deleteBlock(block.id)
        return joinOffset
    }

    /**
     * Create the welcome note. The caller decides whether it is wanted; this
     * only knows how to write it.
     */
    suspend fun seedWelcomeNote(backupAvailable: Boolean): String {
        val noteId = createNote(title = WelcomeNote.TITLE)

        // createNote leaves one empty block for the cursor; the first line takes
        // it over rather than sitting under a blank.
        val firstBlockId = db.blocks().idsForNote(noteId).firstOrNull()
        var previousKey: String? = null

        WelcomeNote.lines(backupAvailable).forEachIndexed { index, line ->
            val orderKey = FractionalIndex.between(previousKey, null)
            val block = Block(
                id = if (index == 0 && firstBlockId != null) firstBlockId else UUID.randomUUID().toString(),
                noteId = noteId,
                orderKey = orderKey,
                type = line.type,
                content = BlockContent(text = line.text),
                hlc = coordinator.engine.stamp(),
            )
            write(block.asChange())
            previousKey = orderKey
        }
        return noteId
    }

    suspend fun update(block: Block) {
        write(block.copy(hlc = coordinator.engine.stamp()).asChange())
    }

    /**
     * Change one block's type, leaving its text alone.
     *
     * The row is re-read here rather than taken from the caller. A caller holds
     * a snapshot from the last time the screen composed, and a block is written
     * whole — so when the formatting bar is tapped a keystroke's write may still
     * be in flight, and a whole-row write built on that snapshot puts the older
     * text back under a newer clock. The character is then gone locally and in
     * the backup. Launches on the same dispatcher run in order, so by the time
     * this reads, the keystroke has landed.
     */
    suspend fun setBlockType(blockId: String, type: BlockType) {
        val current = blockById(blockId) ?: return
        update(current.copy(type = type))
    }

    /** The stored row as the domain type, read fresh. */
    suspend fun blockById(blockId: String): Block? {
        val row = db.blocks().byId(blockId) ?: return null
        val blockType = BlockType.fromWire(row.type) ?: return null
        return Block(
            id = row.id,
            noteId = row.noteId,
            parentId = row.parentId,
            orderKey = row.orderKey,
            type = blockType,
            content = BlockContent.decode(row.content),
            hlc = row.hlc,
            deleted = row.deleted,
        )
    }

    suspend fun deleteBlock(blockId: String) {
        write(Change(Entity.BLOCK, blockId, coordinator.engine.stamp(), deleted = true))
    }

    suspend fun saveTheme(spec: ThemeSpec) {
        write(
            Change(
                entity = Entity.THEME,
                id = spec.id,
                hlc = coordinator.engine.stamp(),
                fields = mapOf(
                    "name" to JsonPrimitive(spec.name),
                    "spec" to JsonPrimitive(spec.encoded()),
                ),
            )
        )
    }

    private suspend fun write(change: Change) {
        coordinator.engine.record(change)
        coordinator.refreshPending()
        coordinator.scheduleSync()
    }
}

File: ios/MyNote/Sources/Sync/NoteRepository.swift
import Foundation
import SwiftData
import MyNoteCore

/// Every write the UI makes goes through here.
///
/// One job: stamp the edit with a fresh clock, persist it locally, queue it for
/// sync, and return — all without awaiting the network. Views call these and
/// treat them as instant, because locally they are.
@MainActor
struct NoteRepository {
    let store: SwiftDataStore
    let coordinator: SyncCoordinator

    // MARK: - Notes

    @discardableResult
    func createNote(title: String = "", parentId: String? = nil, after previous: String? = nil) async -> String {
        let id = UUID().uuidString
        let note = Note(
            id: id,
            title: title,
            parentId: parentId,
            orderKey: FractionalIndex.between(previous, nil),
            hlc: await coordinator.engine.stamp()
        )
        await write(note.asChange())

        // A brand-new page opens with one empty paragraph, so the cursor has
        // somewhere to land instead of an empty screen.
        await appendBlock(to: id, after: nil)
        return id
    }

    func rename(noteId: String, title: String, icon: String?, parentId: String?, orderKey: String) async {
        let note = Note(id: noteId, title: title, icon: icon, parentId: parentId,
                        orderKey: orderKey, hlc: await coordinator.engine.stamp())
        await write(note.asChange())
    }

    func deleteNote(_ noteId: String, blockIds: [String]) async {
        let stamp = await coordinator.engine.stamp()
        await write(Change(entity: .note, id: noteId, hlc: stamp, deleted: true))
        // Tombstone the blocks too, or they would linger on other devices.
        for blockId in blockIds {
            await write(Change(entity: .block, id: blockId,
                               hlc: await coordinator.engine.stamp(), deleted: true))
        }
    }

    // MARK: - Blocks

    @discardableResult
    func appendBlock(to noteId: String, after previous: String?, before next: String? = nil,
                     type: BlockType = .paragraph) async -> String {
        let block = Block(
            noteId: noteId,
            orderKey: FractionalIndex.between(previous, next),
            type: type,
            hlc: await coordinator.engine.stamp()
        )
        await write(block.asChange())
        return block.id
    }

    /// Return pressed: the text before the caret stays, the rest becomes a new
    /// block underneath.
    ///
    /// - Returns: the id of the new block, so the caller can move the caret into it.
    @discardableResult
    func splitBlock(_ block: Block, before: String, after: String,
                    nextOrderKey: String?) async -> String {
        var head = block
        head.content.text = before
        await update(block: head)

        let tail = Block(
            noteId: block.noteId,
            orderKey: FractionalIndex.between(block.orderKey, nextOrderKey),
            // A list carries on as a list; a heading does not, because the line
            // after a heading is almost never another heading.
            type: block.type.continuesOnSplit ? block.type : .paragraph,
            content: BlockContent(text: after),
            hlc: await coordinator.engine.stamp()
        )
        await write(tail.asChange())
        return tail.id
    }

    /// A multi-paragraph paste: one block per paragraph.
    ///
    /// Pasting an article used to drop the whole thing into a single block.
    /// A block's text view does not scroll — it grows to fit — so one block
    /// holding thousands of words has to lay every line of it out again on each
    /// keystroke, and the editor stops responding. Splitting on paste is also
    /// simply what a block editor is for: the paragraphs arrive as paragraphs,
    /// each one movable and styleable on its own.
    ///
    /// - Returns: the id of the last block written, and the caret offset within
    ///   it — the point where the pasted text ends and the block's old trailing
    ///   text resumes.
    func insertParagraphs(
        into block: Block, before: String, paragraphs: [String], after: String,
        nextOrderKey: String?
    ) async -> (blockId: String, offset: Int) {
        // A blank line between paragraphs is a separator, not a paragraph.
        var texts = paragraphs.filter { !$0.trimmingCharacters(in: .whitespaces).isEmpty }
        if texts.isEmpty { texts = [""] }

        var head = block
        head.content.text = before + texts[0]
        await update(block: head)

        guard texts.count > 1 else {
            // One paragraph after all: nothing structural, just longer text.
            let offset = head.content.text.count
            if !after.isEmpty {
                head.content.text += after
                await update(block: head)
            }
            return (block.id, offset)
        }

        var previousKey = block.orderKey
        var lastId = block.id
        var lastOffset = head.content.text.count

        for (index, text) in texts.dropFirst().enumerated() {
            let isLast = index == texts.count - 2
            let key = FractionalIndex.between(previousKey, nextOrderKey)
            let body = isLast ? text + after : text
            let paragraph = Block(
                noteId: block.noteId,
                orderKey: key,
                // Pasted prose is prose. Carrying a heading or a code style
                // across every pasted paragraph is never what was meant.
                type: block.type.continuesOnSplit ? block.type : .paragraph,
                content: BlockContent(text: body),
                hlc: await coordinator.engine.stamp()
            )
            await write(paragraph.asChange())
            previousKey = key
            lastId = paragraph.id
            if isLast { lastOffset = text.count }
        }

        return (lastId, lastOffset)
    }

    /// Backspace at the start of a block: fold it into the one above.
    ///
    /// - Returns: the caret offset in the previous block, i.e. the join point.
    @discardableResult
    func mergeIntoPrevious(_ block: Block, previous: Block) async -> Int {
        let joinOffset = previous.content.text.count

        if !block.content.text.isEmpty {
            var merged = previous
            merged.content.text += block.content.text
            await update(block: merged)
        }
        await deleteBlock(block.id)
        return joinOffset
    }

    func update(block: Block) async {
        var updated = block
        updated.hlc = await coordinator.engine.stamp()
        await write(updated.asChange())
    }

    func deleteBlock(_ blockId: String) async {
        await write(Change(entity: .block, id: blockId,
                           hlc: await coordinator.engine.stamp(), deleted: true))
    }

    /// Create the welcome note, once, on a fresh install.
    ///
    /// Guarded on the database being empty rather than on a flag alone: someone
    /// restoring a backup, or reinstalling with notes already in their Drive,
    /// should not find this sitting on top of their own writing.
    @discardableResult
    func seedWelcomeNoteIfNeeded(existingNoteCount: Int) async -> String? {
        let key = "welcome.seeded"
        let defaults = UserDefaults.standard
        guard !defaults.bool(forKey: key), existingNoteCount == 0 else { return nil }
        defaults.set(true, forKey: key)

        let noteId = await createNote(title: WelcomeNote.title)

        // `createNote` leaves one empty block for the cursor; the first line
        // takes it over rather than sitting under a blank.
        let firstBlockId = (try? await store.blockIds(inNote: noteId))?.first
        var previousKey: String? = nil

        for (index, line) in WelcomeNote.lines.enumerated() {
            if index == 0, let firstBlockId {
                let block = Block(id: firstBlockId, noteId: noteId,
                                  orderKey: FractionalIndex.between(nil, nil),
                                  type: line.type,
                                  content: BlockContent(text: line.text),
                                  hlc: await coordinator.engine.stamp())
                await write(block.asChange())
                previousKey = block.orderKey
                continue
            }

            let block = Block(noteId: noteId,
                              orderKey: FractionalIndex.between(previousKey, nil),
                              type: line.type,
                              content: BlockContent(text: line.text),
                              hlc: await coordinator.engine.stamp())
            await write(block.asChange())
            previousKey = block.orderKey
        }
        return noteId
    }

    // MARK: - Themes

    func saveTheme(_ spec: ThemeSpec) async {
        await write(Change(
            entity: .theme, id: spec.id,
            hlc: await coordinator.engine.stamp(), deleted: false,
            fields: ["name": .string(spec.name), "spec": .string(spec.encoded())]
        ))
    }

    private func write(_ change: Change) async {
        do {
            try await store.recordLocal(change)
        } catch {
            // Local persistence failing is serious, but the in-memory model is
            // still correct for this session, so keep the app usable.
            assertionFailure("local write failed: \(error)")
        }
        await coordinator.refreshPending()
        coordinator.scheduleSync()
    }
}
